sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala

```diff
@@ -69,7 +68,10 @@ class QueryExecution(
   // TODO: Move the planner an optimizer into here from SessionState.
   protected def planner = sparkSession.sessionState.planner
 
-  lazy val isLazyAnalysis: Boolean = logical.containsAnyPattern(LAZY_ANALYSIS_EXPRESSION)
+  lazy val isLazyAnalysis: Boolean = {
+    // Only check the main query as we can resolve LazyOuterReference inside subquery expressions.
+    logical.exists(_.expressions.exists(_.exists(_.isInstanceOf[LazyOuterReference])))
```

Comment:
oh, `logical.containsAnyPattern(LAZY_ANALYSIS_EXPRESSION)` will check subqueries?
I also intended to check the main query. Thanks for the correction!

Reply:
nit: it should be `LazyAnalysisExpression`, although currently only `LazyOuterReference`?